deps/ncrypto/ncrypto.cc

```diff
@@ -2044,4 +2197,417 @@ Result<BIOPointer, bool> EVPKeyPointer::writePublicKey(
   return bio;
 }
 
+// ============================================================================
+
+SSLPointer::SSLPointer(SSL* ssl) : ssl_(ssl) {}
+
+SSLPointer::SSLPointer(SSLPointer&& other) noexcept : ssl_(other.release()) {}
+
+SSLPointer& SSLPointer::operator=(SSLPointer&& other) noexcept {
+  if (this == &other) return *this;
+  this->~SSLPointer();
+  return *new (this) SSLPointer(std::move(other));
+}
+
+SSLPointer::~SSLPointer() {
+  reset();
+}
+
+void SSLPointer::reset(SSL* ssl) {
+  ssl_.reset(ssl);
+}
+
+SSL* SSLPointer::release() {
+  return ssl_.release();
+}
+
+SSLPointer SSLPointer::New(const SSLCtxPointer& ctx) {
+  if (!ctx) return {};
+  return SSLPointer(SSL_new(ctx.get()));
+}
+
+void SSLPointer::getCiphers(
+    std::function<void(const std::string_view)> cb) const {
+  if (!ssl_) return;
+  STACK_OF(SSL_CIPHER)* ciphers = SSL_get_ciphers(get());
+
+  // TLSv1.3 ciphers aren't listed by EVP. There are only 5, we could just
+  // document them, but since there are only 5, easier to just add them manually
+  // and not have to explain their absence in the API docs. They are lower-cased
+  // because the docs say they will be.
+  static const char* TLS13_CIPHERS[] = {"tls_aes_256_gcm_sha384",
+                                        "tls_chacha20_poly1305_sha256",
+                                        "tls_aes_128_gcm_sha256",
+                                        "tls_aes_128_ccm_8_sha256",
+                                        "tls_aes_128_ccm_sha256"};
+
+  const int n = sk_SSL_CIPHER_num(ciphers);
+
+  for (int i = 0; i < n; ++i) {
+    const SSL_CIPHER* cipher = sk_SSL_CIPHER_value(ciphers, i);
+    cb(SSL_CIPHER_get_name(cipher));
+  }
+
+  for (unsigned i = 0; i < 5; ++i) {
+    const char* name = TLS13_CIPHERS[i];
+    cb(name);
+  }
+}
+
+bool SSLPointer::setSession(const SSLSessionPointer& session) {
+  if (!session || !ssl_) return false;
+  return SSL_set_session(get(), session.get()) == 1;
+}
+
+bool SSLPointer::setSniContext(const SSLCtxPointer& ctx) const {
+  if (!ctx) return false;
+  auto x509 = ncrypto::X509View::From(ctx);
+  if (!x509) return false;
+  EVP_PKEY* pkey = SSL_CTX_get0_privatekey(ctx.get());
+  STACK_OF(X509) * chain;
```

Comment:
What's the purpose of this? Can you add a comment? Does this have a side effect?

Reply:
What do you mean? `chain` is used immediately below.

Reply:
This line does not assign itself to anything and it's just a multiplication where the result is not used at all? am i missing something?

Reply:
It's not a multiplication. It's a pointer declaration. The format-cpp just happens to put the * by itself for some reason.